jme3-core/src/main/java/com/jme3/material/Material.java

```diff
@@ -990,7 +996,15 @@ public void render(Geometry geom, RenderManager rm) {
 
     public void write(JmeExporter ex) throws IOException {
         OutputCapsule oc = ex.getCapsule(this);
-        oc.write(def.getAssetName(), "material_def", null);
+
+        final String defAssetName = def.getAssetName();
+
+        oc.write(defAssetName, "material_def", null);
+
+        if (defAssetName == null) {
+            oc.write(def, "material_def_embedded", null);

```

Comment:
I am strongly against this.
The material definition is external to the scene graph. By doing this you allow matdefs to be saved along with a scene and be modified inside it, without having a reference j3md. If this had been the intention in the beginning material and material def wouldn't be separate.

Reply:
It needs for some cases when I generate a mat def on runtime, do you know other options of this?

Reply:
This has nothing to do with serialization. Generating a MaterialDef at runtime is just a matter of creating a new MaterialDef.

Reply:
@Nehon ok, how can I store a material with a generated material def? I can't store the generated material def as file... so the material can't use the material def without asset name.

Reply:
just set the assetName yes. that's the way it's intended to work. And save the material def as a j3md file.

Reply:
@Nehon but I can't save j3md as file :(

Reply:
re: "but I can't save j3md as file :("
Why can't you save it as a .j3md file as already discussed?

Reply:
@pspeed42 because I build it on runtime... so I need to save it as "randomvalue".j3md in the asset, but it will be like trash file. I want to avoid it.

Reply:
Why would you save it before the user explicitly save it and provide a name?

Reply:
@Nehon the main difference between your shader node editor and mine is that my editor edits the file like "shader node project file" which includes last preview material, last version of j3md and some additional things. Of course in the worst case, I can do it using some crap hack solution, but I think when some jme objects can be saved/cloned and other can't be, it's unexpected behavior for a user. I understand that you don't like to support saving/cloning of j3md things, but my point of view, all jme objects should implement JmeClonable and Savable interfaces.

Reply:
You don't need it to implement a proper editor. And if you have to hack around in your code to not clutter the core engine, so be it.